pyro/contrib/autoguide/__init__.py

```diff
@@ -601,6 +621,64 @@ def get_posterior(self, *args, **kwargs):
         return iaf_dist.independent(1)
 
 
+class AutoLaplace(AutoContinuous):
+    """
+    Laplace approximation (quadratic approximation) approximates the posterior
+    :math:`log p(z | x)` by a multivariate normal distribution. Under the hood,
+    it uses Delta distributions to construct a MAP guide over the entire latent
+    space. Its covariance is given by the inverse of the hessian of
+    :math:`-\log p(x, z)` at the MAP point of `z`.
+
+    .. note:: Different to :class:`AutoDelta`, :class:`AutoLaplace` does MAP
+        inference in unconstrained space.
+
+    Usage::
+
+        guide = AutoLaplace(model)
+        svi = SVI(model, guide, ...)
+        gaussian_guide = guide.laplace_approximation()
+
+    By default the mean vector is initialized to zero. To change this default behavior
+    the user should call :func:`pyro.param` before beginning inference, e.g.::
+
+        latent_dim = 10
+        pyro.param("auto_loc", torch.randn(latent_dim))
+    """
+
+    def get_posterior(self, *args, **kwargs):
+        """
+        Returns a diagonal Normal posterior distribution transformed by
+        :class:`~pyro.distributions.iaf.InverseAutoregressiveFlow`.
+        """
+        loc = pyro.param("{}_loc".format(self.prefix),
+                         lambda: torch.zeros(self.latent_dim))
```

Comment:
@fritzo I don't know why we use `lambda` in other autoguide classes, so I keep that pattern. Could you let me know the reason?

Reply:
The `lambda` is a minor optimization that [avoids construction of the tensor](https://github.com/uber/pyro/blob/93c4038/pyro/params/param_store.py#L96) except on the first call. Usually we prefer clarity over minor optimizations, but since this is an internal library I opted for the cheaper lazy initialization.